### PR TITLE
Compute content size, padding and border for each node

### LINKED ISSUE #223: Expose `border` calculated value
## What problem does this solve or what need does it fill?

Currently `Taffy` uses `border` as a second `margin`, which is just used to compute inner margin of children. If some crate want to draw borders `Taffy` doesn't tell where the border begins and where it ends, which forces crate users to recalculate `border` by hand.

## What solution would you like?

Expose the computed `border` value, so users can differentiate between `padding` (external), `border` (middle) and `margin` (internal)

## What alternative(s) have you considered?

Computing `border` by hand, which is very error prone.

## Additional context

In fact `border` isn't part of `Flex`, it's more likely part of `CSS` specification which is used to draw fancy borders on documents, but since `Taffy` already supports `border` (in fact [`border-width`](https://www.w3schools.com/cssref/pr_border-width.asp)), it would be a nice feature to expose this to users than can actually draw their borders.

### PR DESCRIPTION
**Pull request by nicoburns**
# Objective

- Closes https://github.com/DioxusLabs/taffy/issues/470
- Closes #223

## Changes made

- `Overflow::Clip` variant added to the `Overflow` enum. Layout with `Clip` behaves the same as `Visible`, except the overflowing contents don't contribute to the ancestor nodes' content sizes.
- `content_size` feature flag added
- `content_size` field added to `Layout` struct (stored on Node) and `LayoutOutput` (returned to parent) structs
- `border` and `padding` fields added to the `Layout` struct
- Implement `content_size` computation for all layout modes
- Add `scroll_width` and `scroll_height` methods to `Layout` and add tests for scroll width/height
- Remove caching run_mode exception for leaf layout. We now need to run leaf layout

## Todo

- [x] Feature flag the block layout implementation
- [x] Feature flag the leaf layout implementation (if appropriate)

## Context

Useful for any UI toolkit wanting to implement scrolling (e.g. https://github.com/bevyengine/bevy/pull/8104). Although it's worth noting that this implementation assumes that every node is a scroll node, and it may be possible to do this cheaper if you can avoid that assumption at a higher level.

## Feedback wanted

- General code review
- Should this be merged as-is (causes 5-20% performance hit).  Or can anyone think of any way to reduce to reduce the performance impact - as far as I can tell is the actually computation of the content sizes from the sizes of the child nodes?
- If it should be merged, should it be enabled by default?

### CODE REVIEW

File: scripts/gentest/src/main.rs [hunk 1]
```diff
@@ -234,11 +234,17 @@ fn generate_assertions(ident: &str, node: &Value, use_rounding: bool) -> TokenSt
...
 
     let read_f32 = |s: &str| layout[s].as_f64().unwrap() as f32;
+    let read_naive_f32 = |s: &str| node["naivelyRoundedLayout"][s].as_f64().unwrap() as f32;
     let width = read_f32("width");
     let height = read_f32("height");
     let x = read_f32("x");
     let y = read_f32("y");
 
+    // TODO: Use clientWidth/clientHeight to take into account scrollbars
+    // TODO: Handle rounding better
```
Comment (TimJentzsch) [inline scripts/gentest/src/main.rs:244]: There are some TODOs here, do they need to be resolved in this PR or is this for a different issue?

File: scripts/gentest/test_helper.js [hunk 2]
```diff
@@ -8,6 +8,22 @@ function getScrollBarWidth() {
...
+//   for (node of e.childNodes) {
+//     let boundingRect = e.getBoundingClientRect();
+//     let style = getComputedStyle(e);
+//     min_x = min_x.min(boundingRect.left);
+//     max_x = max_x.max(boundingRect.right);
+//     min_y = min_y.min(boundingRect.top);
+//     max_y = max_y.min(boundingRect.bottom);
+//   }
+//   let padding = getComputedStyle(e)
+// }
```
Comment (TimJentzsch) [inline scripts/gentest/test_helper.js:25]: Let's remove this if we don't need it anymore. We can use git to pull it back if needed

File: src/compute/flexbox.rs [hunk 3]
```diff
@@ -1733,24 +1737,43 @@ fn calculate_flex_item(
...
-        order: item.order,
-        size: preliminary_size_and_baselines.size,
-        location: Point {
-            x: if direction.is_row() { offset_main } else { offset_cross },
-            y: if direction.is_column() { offset_main } else { offset_cross },
-        },
+    let location = Point {
+        x: if direction.is_row() { offset_main } else { offset_cross },
+        y: if direction.is_column() { offset_main } else { offset_cross },
     };
```
Comment (TimJentzsch) [inline src/compute/flexbox.rs:1751]: Perhaps easier and only needs one check (will hardly make a perf difference though):
```suggestion
    let location = if direction.is_row() {
        Point {
            x: offset_main,
            y: offset_cross,
        }
    } else {
        Point {
            x: offset_cross,
            y: offset_main,
        }
    };
```

File: src/compute/flexbox.rs [hunk 4]
```diff
@@ -1733,24 +1737,43 @@ fn calculate_flex_item(
...
+            width: match item.overflow.x {
+                Overflow::Visible => f32_max(size.width, content_size.width),
+                _ => size.width,
+            },
+            height: match item.overflow.y {
+                Overflow::Visible => f32_max(size.height, content_size.height),
+                _ => size.height,
+            },
+        };
+        if size_content_size_contribution.width > 0.0 && size_content_size_contribution.height > 0.0 {
```
Comment (TimJentzsch) [inline src/compute/flexbox.rs:1769]: I can see this > 0.0 check 3 times in this PR, perhaps it warrants a helper function on `Size`

File: src/compute/flexbox.rs [hunk 5]
```diff
@@ -1800,15 +1828,24 @@ fn calculate_layout_line(
...
 #[inline]
-fn final_layout_pass(tree: &mut impl PartialLayoutTree, flex_lines: &mut [FlexLine], constants: &AlgoConstants) {
+fn final_layout_pass(
+    tree: &mut impl PartialLayoutTree,
+    flex_lines: &mut [FlexLine],
+    constants: &AlgoConstants,
+) -> Size<f32> {
     let mut total_offset_cross = constants.content_box_inset.cross_start(constants.dir);
 
+    #[cfg_attr(not(feature = "content_size"), allow(unused_mut))]
```
Comment (TimJentzsch) [inline src/compute/flexbox.rs:1838]: Why is the unused mut fine here?

File: src/compute/flexbox.rs [hunk 6]
```diff
@@ -2030,15 +2077,36 @@ fn perform_absolute_layout_on_absolute_children(
...
-            order: order as u32,
-            size: final_size,
-            location: Point {
-                x: if constants.is_row { offset_main } else { offset_cross },
-                y: if constants.is_column { offset_main } else { offset_cross },
-            },
+        let location = Point {
+            x: if constants.is_row { offset_main } else { offset_cross },
+            y: if constants.is_column { offset_main } else { offset_cross },
         };
```
Comment (TimJentzsch) [inline src/compute/flexbox.rs:2083]: Nit: Can probably also be reduced to a single if as my other comment above

File: src/compute/taffy_tree.rs
Comment (TimJentzsch) [inline src/compute/taffy_tree.rs:1]: At first sight, this looks unrelated to the goal of this PR, can this be pulled into a different PR for easier reviewing?

**Review by TimJentzsch — COMMENTED** [review]
I have to admit, it's becoming quite hard for me to grasp all the layout logic as it's becoming more and more complex.
I wonder if we can improve this somehow, maybe with more inline docs or with better abstractions or if the complexity of Flexbox and the performance considerations make this a necessity.

Regarding the performance, it would of course be great if we only had to pay this cost if scrolling is enabled, but I assume this would be difficult to implement right now.
I assume Bevy and other clients will always have this feature enabled, so it would be great to reduce the performance cost somehow.
If the feature is enabled by default or not is less important IMO.

File: src/compute/taffy_tree.rs
Comment (nicoburns) [reply src/compute/taffy_tree.rs:1]: Ah, this one's a merge conflict. This file was deleted after this branch was created and seems to have been recreated when rebasing.

File: src/compute/flexbox.rs (hunk 5 above)
Comment (nicoburns) [reply src/compute/flexbox.rs:1838]: Well, unused `mut` is harmless in general. Except that it's confusing for readers of the code. In this case the variable is mutable when the feature is enabled and immutable otherwise. My judgement here was that it was more readable to do this than conditionally define the variable depending on the feature flag which would look like:

```rust 
#[cfg(feature = "content_size")]
let mut content_size = Size::ZERO;
#[cfg(not(feature = "content_size"))]
let content_size = Size::ZERO;
```

File: scripts/gentest/src/main.rs (hunk 1 above)
Comment (nicoburns) [reply scripts/gentest/src/main.rs:244]: Ah yes, that is to take into account the affect that scrollbars have on the scroll width/height (they take up space, so they actually make you scroll slightly further). That could either be done as part of this PR, or as a follow up PR.

File: src/compute/flexbox.rs (hunk 3 above)
Comment (nicoburns) [reply src/compute/flexbox.rs]: I've gone with
```rust
let location = match constants.is_row {
      true => Point { x: offset_main, y: offset_cross },
      false => Point { x: offset_cross, y: offset_main },
};
```
to keep it short :)

File: scripts/gentest/src/main.rs [hunk 7]
```diff
@@ -253,21 +259,31 @@ fn generate_assertions(ident: &str, node: &Value, use_rounding: bool) -> TokenSt
...
         quote!(
-            let Layout { size, location, .. } = taffy.layout(#ident).unwrap();
+            #[cfg_attr(not(feature = "content_size"), allow(unused_variables))]
+            let layout @ Layout { size, location, .. } = taffy.layout(#ident).unwrap();
             assert!(size.width - #width < 0.1, "width of node {:?}. Expected {}. Actual {}", #ident,  #width, size.width);
             assert!(size.height - #height < 0.1, "height of node {:?}. Expected {}. Actual {}", #ident,  #height, size.height);
             assert!(location.x - #x < 0.1, "x of node {:?}. Expected {}. Actual {}", #ident,  #x, location.x);
             assert!(location.y - #y < 0.1, "y of node {:?}. Expected {}. Actual {}", #ident,  #y, location.y);
+            #[cfg(feature = "content_size")]
+            assert!(layout.scroll_width() - #scroll_width < 0.1, "scroll_width of node {:?}. Expected {}. Actual {}", #ident,  #scroll_width, layout.scroll_width());
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply scripts/gentest/src/main.rs:284]: It would also be nice to pull out 0.1 into a shared epsilon value for clarity.

File: src/style/mod.rs [hunk 8]
```diff
@@ -132,12 +132,18 @@ impl Default for Position {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum Overflow {
     /// The automatic minimum size of this node as a flexbox/grid item should be based on the size of it's content.
+    /// Content that overflows this node *should* contribute to the scroll region of it's parent.
```
Comment (alice-i-cecile) [inline src/style/mod.rs:135]: Nit: "its" should be used rather than "it's" throughout the docs for this type.

File: scripts/gentest/src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicoburns) [reply scripts/gentest/src/main.rs:284]: Hmm... yeah this almost certainly ought to be comparing absolute difference. For the other values above too. I guess this wasn't noticed because it's working in practice (I think most tests are using the rounded values anyway, it's only a handful that are using the unrounded values). I believe the scroll values might able to use exact equality in both cases because we don't have access to the the float values of the scroll width/height anyway.

File: src/style/mod.rs (hunk 8 above)
Comment (nicoburns) [reply src/style/mod.rs:135]: Gah. I do this every single time (because my intuition is that both `it's` and `its` ought to have an apostrophe).

**Review by alice-i-cecile — APPROVED** [review]
> Should this be merged as-is (causes 5-20% performance hit). Or can anyone think of any way to reduce to reduce the performance impact - as far as I can tell is the actually computation of the content sizes from the sizes of the child nodes?

I'm okay to merge as is: I think this is fundamentally a necessary feature for the spec.

> If it should be merged, should it be enabled by default?

Hmm. So ultimately our users are UI crates that consume taffy. Some of them will be simple, and won't support scrolling, others are more complex and do. Unless they make a feature flag to mirror this, that control won't be exposed to users.

In Bevy for example, I would probably push for a mirror of this feature flag, and then enable it by default in Bevy.

On the balance, I think we should keep this on by default, to match our existing policy of enabling default features for multiple different layout algorithms, even if not all of them are useful to all UI crates.

File: scripts/gentest/src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicoburns) [reply scripts/gentest/src/main.rs]: These have been updated to use `.abs()`.

File: scripts/gentest/src/main.rs (hunk 1 above)
Comment (nicoburns) [reply scripts/gentest/src/main.rs]: This has now been implemented

**Comment by nicoburns** [comment]
There's still a performance hit with this, but it's much better since #579 (now more like 5-8%). So I think I'm going to enable this by default. I suspect we may also need to tweak the computed size in future (as we discover bugs), but I think this is good enough as a first cut at the feature.
